webhook_server_container/libs/github_api.py

```diff
@@ -2071,3 +2086,95 @@ def run_podman_command(self, command: str, pipe: bool = False) -> Tuple[bool, st
             return run_command(command=command, log_prefix=self.log_prefix, pipe=pipe)
 
         return rc, out, err
+
+    def get_approvers_and_reviewers(self) -> dict[str, dict[str, list[str]]]:
+        # Dictionary mapping OWNERS file paths to their approvers and reviewers
+        _owners: dict[str, dict[str, list[str]]] = {}
+
+        max_owners_files = 1000  # Configurable limit
+        owners_count = 0
+
+        tree = self.repository.get_git_tree(self.pull_request_branch, recursive=True)
+        for element in tree.tree:
+            if element.type == "blob" and element.path.endswith("OWNERS"):
+                owners_count += 1
+                if owners_count > max_owners_files:
+                    self.logger.error(f"{self.log_prefix} Too many OWNERS files (>{max_owners_files})")
+                    break
+
+                content_path = element.path
+                _path = self.repository.get_contents(content_path)
+                if isinstance(_path, list):
+                    _path = _path[0]
+
+                try:
+                    content = yaml.safe_load(_path.decoded_content)
+                    if self._validate_owners_content(content, content_path):
+                        # Use Path for consistent path handling
+                        parent_path = str(Path(content_path).parent)
+                        if not parent_path:
+                            parent_path = "."
+                        _owners[parent_path] = content
+
+                except yaml.YAMLError as exp:
+                    self.logger.error(f"{self.log_prefix} Invalid OWNERS file {content_path}: {exp}")
+                    continue
+
+        self.logger.debug(f"{self.log_prefix} Owners file mapping: {_owners}")
+        return _owners
+
+    def get_all_approvers(self) -> list[str]:
+        _approvers: list[str] = []
+        for list_of_approvers in self.owners_data_for_changed_files()["approvers"]:
+            for _approver in list_of_approvers:
+                _approvers.append(_approver)
+
+        return list(set(self.root_approvers + _approvers))
+
+    def get_all_reviewers(self) -> list[str]:
+        _reviewers: list[str] = []
+        for list_of_reviewers in self.owners_data_for_changed_files()["reviewers"]:
+            for _approver in list_of_reviewers:
+                _reviewers.append(_approver)
+
+        return list(set(self.root_reviewers + _reviewers))
+
+    def owners_data_for_changed_files(self) -> dict[str, list[list[str]]]:
+        data: dict[str, list[list[str]]] = {"approvers": [], "reviewers": []}
+
+        changed_folders = {Path(cf).parent for cf in self.changed_files}
+
+        for changed_folder_path in changed_folders:
```

Comment:
I am a bit confused here. Looking at this code, if this is the folder structure:
```
-root (owner0, approver 0 and reviewer 0) 
         -|-folder1(owner1 (approver 1 and reviewer 1)) - file1, file2, file3
          | -folder2(owner2 approver 2 and reviewer2)) - file 4, file 5)
          |-folder 3
```
if a patch touches folder 1 and folder 3, then the reviewers and approvers would be approver 0, approver 1, reviewer 0, reviewer1 is this assumption right?
While for reviewers everyone should be included, for approvers, I would need more restrictive approach. e.g. if the any of the changed_folder_path, is not fully under a SIG's ownership, the approval should be based on OWNER file under the first common ancestor. I think if we can talk I can explain it better.